runtime/src/main/java/org/corfudb/runtime/view/ManagementView.java

```diff
@@ -221,6 +226,86 @@ private ClusterStatus getClusterHealth(Layout layout, Set<String> peerResponsive
         return peerResponsiveNodes;
     }
 
+    /**
+     * Fetches the initial layout to fetch the cluster status.
+     * This does not consume the corfu runtime layout as that can get stuck in an indefinite retry
+     * loop.
+     *
+     * @param layoutServers Layout Servers to fetch the layout from.
+     * @return Latest layout from the servers or null if none of them responded with a layout.
+     */
+    private Layout getClusterStatusLayout(List<String> layoutServers) {
+        Layout layout = null;
+        Map<String, CompletableFuture<Layout>> layoutFuturesMap = new HashMap<>();
+        // Get layout futures for layout requests to all layout servers.
+        for (String server : layoutServers) {
+            try {
+                IClientRouter router = runtime.getRouter(server);
+                layoutFuturesMap.put(server,
+                        new LayoutClient(router, Layout.INVALID_EPOCH).getLayout());
+            } catch (Exception e) {
+                log.error("getClusterStatus: Exception encountered connecting to {}. ",
+                        server, e);
+                CompletableFuture<Layout> cf = new CompletableFuture<>();
+                cf.completeExceptionally(e);
+                layoutFuturesMap.put(server, cf);
+            }
+        }
+
+        // Wait on the Completable futures and retain the layout with the highest epoch.
+        for (Entry<String, CompletableFuture<Layout>> entry : layoutFuturesMap.entrySet()) {
+            try {
+                Layout nodeLayout = entry.getValue().get();
+
+                if (layout == null || nodeLayout.getEpoch() > layout.getEpoch()) {
+                    layout = nodeLayout;
+                }
+            } catch (InterruptedException ie) {
+                throw new UnrecoverableCorfuInterruptedError(ie);
+            } catch (ExecutionException ee) {
+                log.error("getClusterStatus: Error while fetching layout from {}.",
+                        entry.getKey(), ee);
+            }
+        }
+        return layout;
+    }
+
+    /**
+     * Create node status map.
+     * Assigns a node status to all the nodes in the layout.
+     * UP: If the node is responsive and present in all the segments OR present in none (for cases
+     * where a node is a Layout or Sequencer Server only and does not possess a LogUnit component.)
+     * DB_SYNCING: If the node is present in a few segments but not all. This node is syncing data.
+     * DOWN: If the node is unresponsive.
+     *
+     * @param layout            Layout used to compute cluster status
+     * @param responsiveServers Set of responsive servers excluding unresponsive Servers list.
+     * @param allServers        All servers in the layout.
+     * @return Map of nodes mapped to their status.
+     */
+    private Map<String, NodeStatus> createNodeStatusMap(Layout layout,
```

Comment:
I have a concern/question which doesn't need to be addressed on this PR but we might need to think about it at some point.
The methods added to ManagementView (along with some previous code) are responsible for evaluating and instantiating a ClusterStatusReport. What do you think about making the report immutable AND moving responsibility of evaluating the status to ClusterStatusReport? This way, we have two benefits of removing from ManagementView the burden of creating the reports as well as side benefits of "immutable" reports.

Reply:
According to me, a report should be a static object and should not be responsible for evaluating  (heartbeating and creating the status map) this report. Let me know if I answered your question.